fix: add Chromium Web Browser to the transition list

isantop opened the pull request:
The Chromium flatpak on Flathub is more up to date than the debian version and doesn't cause any major issues. Therefore we are transitioning it to Flatpak.

Commit:
fix: add Chromium Web Browser to the transition list.

leviport reviewed (approved):
Works like a charm. I can still see the .deb version in Pop Shop, so we'll probably want to take that down eventually as well.

leviport reviewed (changes requested):
Actually hang on. It looks like none of the bookmarks, history, or preferences from the .deb install moved over to the flatpak install. Users that haven't linked it to a google account might lose that info, which many would find upsetting. I think we need to think about that a bit.

isantop commented:
That's a good point. Furthermore, a lot of that literally can't be exported (for example, saved passwords).

WatchMkr commented:
Does the copy in the transition app sufficiently cover that user preferences do not transfer? This doesn't seem different than a password manager, except that there may be many more users.

The transition has to occur at some point. Old browsers are a significant security hole.

ids1024 commented:
> This doesn't seem different than a password manager, except that there may be many more users.

A browser profile is also more complicated, so there's possibly more that could go wrong. But I know Firefox profiles are just stored in a particular directory (I believe I once copied one from Window to Linux). It is presumably similar for Chromium.

Though if there's any possible situation where this could transition to a older Chromium version (if Flathub had an older version than the Ubuntu repos), moving a profile to an older version is likely problematic.

I think this could work, but a degree of caution is warranted.

isantop commented:
This is the current copy we're using. IMO it doesn't really explain what might happen. I also wonder if it's a good idea to get extremely application specific with these instructions.

![Screenshot from 2021-01-07 09-33-57](https://user-images.githubusercontent.com/5883565/103918585-36ed3d80-50cc-11eb-948d-8000cad439f3.png)

![Screenshot from 2021-01-07 09-35-36](https://user-images.githubusercontent.com/5883565/103918597-3a80c480-50cc-11eb-94fe-cfa9dc740799.png)


On the other hand, removing Chromium does not remove the data folder from the user's home folder, so if the user does remove the Chromium deb and then later needs access to data that was present there, they can re-install the deb and the data will be there still.

isantop commented:
@mmstick Can we leave the Chromium package in our repository (or possibly replace it with a dummy package) that can be pinned to avoid a user installing a newer version of Chromium from Apt than the one we're presently shipping with? If so, that should handle the caution case that @ids1024 mentioned (presuming that the user hasn't added a PPA that has a newer one available; the official Chromium PPA hasn't been updated since Chromium 75 in 2019: https://launchpad.net/~chromium-team/+archive/ubuntu/stable )

If we can do this, I can add a simple directory copy configuration migration; Chromium should successfully migrate any profile from an old version to whatever is current.

mmstick commented:
We can make it a dummy package

isantop commented:
As long as we don't remove the browser for current users (at least initially), that should work well. We'll need to remove the Appdata so that it doesn't appear in AppCenter.

I'll get to work updating this PR with the configuration migration. Since this is a fairly simple case, that shouldn't be too bad.

Commit:
feat(package): add simple configuration migration

Adds simple configuration migration to the transition app. This requires that
the configuration be identical between two application versions (or the
application can migrate them), that both the old and new configuration are
stored in standardized locations and don't change based on other configuration,
that the old and new configuration are both stored in a directory located
somewhere within the user's home directory, and that copying the contents of
the old directory (and their attributes) to the new directory will preserve
the configiuration. Use cases outside this are currently out-of-scope and
should be handled manually by the user.

Also adds such configuration migration to Chromium.

https://github.com/pop-os/transition/pull/13#pullrequestreview-562990340

Commit:
fix(package): fix nonetype error

Commit:
fix(package): need to expand home folder to find the correct path

Commit:
fix(application): migrate configs after installation, rather than before removal

Commit:
fix(package): fix typo

ids1024 commented:
> the official Chromium PPA hasn't been updated since Chromium 75 in 2019

I suppose Ubuntu's Chromium team is no longer really maintaining that since they have switched to snap.

Though interestingly their dev/beta PPAs are updated today (perhaps automatically):
* https://launchpad.net/~chromium-team/+archive/ubuntu/dev
* https://launchpad.net/~chromium-team/+archive/ubuntu/beta

Edit: Though actually, looking at those PPAs, they seem to build only for Bionic and Xenial.

Commit:
fix(package): ensure we're using the correct variables

isantop commented:
The latest changes add infrastructure for simple configuration migration to the transition app. This requires that the configuration be identical between two application versions (or the application can migrate them), that both the old and new configuration are stored in standardized locations and don't change based on other configuration, that the old and new configuration are both stored in a directory located somewhere within the user's home directory, and that copying the contents of the old directory (and their attributes) to the new directory will preserve the configuration. 

If there is already a configuration directory for the Flatpak version, it silently fails to perform the migration (preventing any data there from being overwritten). I think this is a niche enough corner case that we probably don't need to worry too much about this; users who already set up the Flatpak version are unlikely to be using both the Apt and Flatpak versions anyway, as there's not really any benefit to doing so.

It also adds the configuration migration process for Chromium. I have tested this and the migration appears to be successful; Opening the Flatpak version of Chromium present the exact session I had open in the Apt version. We should test that this works for multiple people, not just me, and additionally perform some regression testing with other apps which don't do any configuration migration (to ensure that this process is correctly skipped for apps which don't use it).

jacobgkau commented:
The transition (including config migration) is working on Groovy. On a system with multiple users, the Flatpak was only installed by default on the user that ran the transition, but after installing the Flatpak from the Pop!_Shop on the remaining user, it still had its old data.

However, this is not working on Focal (and this PR is merging into `master_focal`.) I'm currently unable to install the Chromium flatpak on Focal from the Transition app, the Pop!_Shop, or the terminal.

![Screenshot from 2021-01-11 10-09-48](https://user-images.githubusercontent.com/7199422/104215026-7205ae80-53f5-11eb-88c4-36a7510cc4d6.png)

The terminal says:

```
Error: org.chromium.Chromium needs a later flatpak version
error: Failed to install org.chromium.Chromium: app/org.chromium.Chromium/x86_64/stable needs a later flatpak version (1.8.2)
```

It looks like Focal is shipping Flatpak `1.6.5-0ubuntu0.1`, while Groovy is shipping `1.8.2-1`. It seems like we would either need to ship a newer version of Flatpak for Focal or not make this transition on Focal.

ids1024 commented:
> It seems like we would either need to ship a newer version of Flatpak for Focal or not make this transition on Focal.

I think regardless of this transition, we want users to be able to use new Flatpaks on Focal (being able to install new software on an LTS is one of the potential advantages of Flatpak).

So if it's not hugely unusual for Flatpaks to require a newer `flatpak` version, and it works fine with a backport, we should do that.

mmstick commented:
Backporting Flatpak is perfectly fine if we can do so.

ids1024 commented:
For comparison, `buster-backports` already has 1.8.4: https://packages.debian.org/buster-backports/flatpak

But I suppose Flatpak is (understandably) less of a priority for Ubuntu because of snap.

I'll try creating a repo/branch with the packaging from Debian. I doubt it will pose much of an issue.

ids1024 commented:
https://github.com/pop-os/flatpak/tree/flatpak-1.8.4 has now built successfully on CI for both focal and groovy.

mmstick commented:
The flatpak package looks good to me.

jacobgkau commented:
The new Flatpak package does allow installing the Chromium flatpak on Focal, and doesn't cause any issues on Groovy.

Testing the configuration migration some more, it actually looks like the configuration only gets transferred for the current user (who's running the transition app), not for other users who were using the deb package. This isn't actually destructive, since the files are still present at `.config/Chromium` and can be copied manually to `.var/app/org.chromium.Chromium/config/chromium` (doing this manually requires removing `.var/app/org.chromium.Chromium/config/chromium` if it exists before copying, otherwise the copied data doesn't seem to show up.)

Are we okay with the configuration only being transferred for the current user, or would it be possible to try to transfer configuration for each user present?

jacobgkau commented:
Based on yesterday's meeting, the migration of configuration for only the current user is acceptable (config for other users is not removed and can be migrated manually, the copy already suggests backing up/exporting data before uninstalling the deb, and the transition is able to be run at the users' convenience.)

However, I am having one other problem: if Android Studio (the only app above Chromium alphabetically) is transitioned, then the config is not copied. This does not occur with Chromium and other apps, just Chromium and Android Studio. Terminal output:

```
system76@pop-os:~$ pop-transition 
showing window
Install clicked
Installing Flatpaks...
Installing Android Studio flatpak com.google.AndroidStudio.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pop_transition/application.py", line 127, in on_install_clicked
    if package.old_config and package.new_config:
  File "/usr/lib/python3/dist-packages/pop_transition/package.py", line 234, in old_config
    return Path.home() / self._old_config
TypeError: unsupported operand type(s) for /: 'PosixPath' and 'NoneType'
Installing Chromium flatpak org.chromium.Chromium.
Looking for ID from ref runtime/com.google.AndroidStudio.Locale/x86_64/stable
/usr/lib/python3/dist-packages/pop_transition/flatpak.py:107: Warning: g_value_get_flags: assertion 'G_VALUE_HOLDS_FLAGS (value)' failed
  transaction.run()
Looking for ID from ref runtime/com.google.AndroidStudio.Locale/x86_64/stable
Looking for ID from ref runtime/org.chromium.Chromium.Codecs/x86_64/stable
Looking for ID from ref runtime/org.chromium.Chromium.Codecs/x86_64/stable
Looking for ID from ref runtime/org.chromium.Chromium.Locale/x86_64/stable
Looking for ID from ref runtime/org.chromium.Chromium.Locale/x86_64/stable
Looking for ID from ref app/org.chromium.Chromium/x86_64/stable
Found ID org.chromium.Chromium in package Chromium
Looking for ID from ref app/org.chromium.Chromium/x86_64/stable
Found ID org.chromium.Chromium in package Chromium
Looking for ID from ref app/com.google.AndroidStudio/x86_64/stable
Found ID com.google.AndroidStudio in package Android Studio
Looking for ID from ref app/com.google.AndroidStudio/x86_64/stable
Found ID com.google.AndroidStudio in package Android Studio
```

It seems like once config migration errors out for Android Studio (since the variables are set to `None`), it's not attempted for Chromium.

Commit:
fix(application): catch FileExistsError and print a more helpful message

Commit:
fix(package): return None if there is no old config to migrate

Otherwise things will error out and prevent items from being migrated when they should have been.

Fixes issue in https://github.com/pop-os/transition/pull/13#issuecomment-761147626

jacobgkau reviewed (approved):
Config migration is now working even when Android Studio is also being transitioned. Not seeing any other issues or topics that haven't already been discussed and accepted.

ids1024 commented:
Since this merged into `master_focal`, I suppose it also needs to be merged into `master_groovy`?
